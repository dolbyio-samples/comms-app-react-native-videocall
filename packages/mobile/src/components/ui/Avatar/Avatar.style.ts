import { StyleSheet } from 'react-native';

export default StyleSheet.create({
  wrapper: {
    display: 'flex',
    borderWidth: 2,
    justifyContent: 'center',
    alignItems: 'center',
  },
});
