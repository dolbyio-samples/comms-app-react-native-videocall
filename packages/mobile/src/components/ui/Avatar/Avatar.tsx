/* eslint-disable react/jsx-props-no-spreading */
import type { Participant } from '@dolbyio/comms-sdk-react-native/lib/typescript/services/conference/models';
import type { ColorKey, Sizes } from '@dolbyio/comms-uikit-common';
import React, { useMemo } from 'react';
import { View, ViewProps } from 'react-native';

import useTheme from '../../../hooks/useTheme';
import { stringToNumber } from '../../../utils/stringToNumber.util';
import Text, { TextType } from '../Text/Text';

import styles from './Avatar.style';

type AvatarSizes = Extract<Sizes, 'xs' | 's' | 'm' | 'l'>;

export type AvatarProps = ViewProps & {
  participant: Participant | string | undefined;
  size?: AvatarSizes;
  borderColor?: ColorKey;
  testID?: string;
};

const AvatarSizesMap: { [key in AvatarSizes]: { width: number; height: number; text: TextType } } = {
  xs: { width: 24, height: 24, text: 'avatarXS' },
  s: { width: 40, height: 40, text: 'avatarS' },
  m: { width: 48, height: 48, text: 'avatarM' },
  l: { width: 80, height: 80, text: 'avatarL' },
};

const Avatar = ({ participant, size = 'l', borderColor, testID, ...props }: AvatarProps) => {
  if (participant === undefined) {
    return null;
  }
  const { avatars, getColor } = useTheme();
  const name = typeof participant === 'string' ? participant : participant.info.name;

  const [firstNameLetter, lastNameLetter] = useMemo(() => {
    let firstValue;
    let lastValue;
    if (name) {
      firstValue = name.charAt(0);
      const nameArray = name.split(' ');
      lastValue = nameArray.length > 1 ? nameArray[nameArray.length - 1].charAt(0) : '';
    }
    return [firstValue, lastValue];
  }, [name]);

  // Generate random background color
  const backgroundColor = useMemo(() => avatars[stringToNumber(name) % avatars.length], [name]);
  const { width, height, text } = AvatarSizesMap[size];
  const borderRadius = width / 2;
  return (
    <View
      testID={testID}
      style={[
        styles.wrapper,
        { width, height, borderRadius, backgroundColor, borderColor: getColor(borderColor, 'white') },
      ]}
      {...props}
    >
      <Text testID={`${testID}-text`} type={text}>
        {firstNameLetter?.toUpperCase()}
        {['m', 'l'].includes(size) ? lastNameLetter?.toUpperCase() : undefined}
      </Text>
    </View>
  );
};

export default Avatar;
