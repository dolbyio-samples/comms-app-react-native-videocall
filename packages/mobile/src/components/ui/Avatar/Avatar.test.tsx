import type { Participant } from '@dolbyio/comms-sdk-react-native/lib/typescript/services/conference/models';
import { defaultTheme as theme } from '@dolbyio/comms-uikit-common';
import React from 'react';
import { Text } from 'react-native';
// eslint-disable-next-line import/no-extraneous-dependencies
import renderer from 'react-test-renderer';

import { render, mergeStyles } from '../../../utils/test-utils';

import Avatar from './Avatar';

const { colors } = theme;

const testID = 'testID';
const textElmTestID = `${testID}-text`;
const borderColor = colors.purple[400];
const confUsr1 = 'FOO';
const confUsr2 = 'UIKIT QA';
const testParticipant: Participant = {
  id: 'testParticipant_id',
  info: {
    avatarUrl: 'http://dummy.avatar.org/',
    externalId: 'testParticipant_external_id',
    name: 'dolby sydney',
  },
  status: 'INACTIVE',
  streams: [],
  type: 'listener',
};

const commonExpStyle = {
  display: 'flex',
  borderWidth: 2,
  justifyContent: 'center',
  alignItems: 'center',
  backgroundColor: '#FAC819',
  borderColor: 'white',
};

const commonTextElmExpStyle = {
  letterSpacing: 1.22,
  fontWeight: '600',
  fontFamily: 'Roboto',
  color: '#FFFFFF',
  textAlign: 'left',
};

describe('Avatar component', () => {
  test('Passes TestID', () => {
    const { getByTestId } = render(<Avatar participant={testParticipant} borderColor={borderColor} testID={testID} />);

    const element = getByTestId(testID);
    expect(element).not.toBeNull();

    const txtElement = getByTestId(textElmTestID);
    expect(txtElement).not.toBeNull();

    const textElmList = element.findAllByType(Text);
    expect(textElmList).toHaveLength(1);
  });
  test('should return null', () => {
    const { getByTestId } = render(<Avatar borderColor={borderColor} testID={testID} />);
    expect(() => {
      getByTestId(testID);
    }).toThrow('Unable to find an element with testID: testID');
  });
  test('matches style with default size', () => {
    const expStyles = {
      ...commonExpStyle,
      height: 80,
      width: 80,
      borderRadius: 40,
    };
    const textElmExpStyle = {
      ...commonTextElmExpStyle,
      fontSize: 32,
      lineHeight: 40,
    };
    const { getByTestId } = render(<Avatar participant={testParticipant} testID={testID} />);
    const element = getByTestId(testID);
    const styles = mergeStyles(element.props.style);
    const txtElement = getByTestId(textElmTestID);
    const textElmStyle = mergeStyles(txtElement.props.style);
    expect(styles).toMatchObject(expStyles);
    expect(textElmStyle).toMatchObject(textElmExpStyle);
  });

  test('matches style with size:xs', () => {
    const expStyles = {
      ...commonExpStyle,
      height: 24,
      width: 24,
      borderRadius: 12,
    };
    const textElmExpStyle = {
      ...commonTextElmExpStyle,
      fontSize: 10,
      lineHeight: 18,
    };

    const size = 'xs';
    const { getByTestId } = render(
      <Avatar participant={testParticipant} size={size} borderColor={borderColor} testID={testID} />,
    );
    const element = getByTestId(testID);
    const styles = mergeStyles(element.props.style);
    const txtElement = getByTestId(textElmTestID);
    const textElmStyle = mergeStyles(txtElement.props.style);
    expect(styles).toMatchObject(expStyles);
    expect(textElmStyle).toMatchObject(textElmExpStyle);
  });
  test('matches style with size:s', () => {
    const expStyles = {
      ...commonExpStyle,
      height: 40,
      width: 40,
      borderRadius: 20,
    };
    const textElmExpStyle = {
      ...commonTextElmExpStyle,
      fontSize: 16.6667,
      lineHeight: 23,
    };
    const size = 's';
    const { getByTestId } = render(
      <Avatar participant={testParticipant} size={size} borderColor={borderColor} testID={testID} />,
    );
    const element = getByTestId(testID);
    const styles = mergeStyles(element.props.style);
    const txtElement = getByTestId(textElmTestID);
    const textElmStyle = mergeStyles(txtElement.props.style);
    expect(styles).toMatchObject(expStyles);
    expect(textElmStyle).toMatchObject(textElmExpStyle);
  });
  test('matches style with size:m', () => {
    const expStyles = {
      ...commonExpStyle,
      height: 48,
      width: 48,
      borderRadius: 24,
    };
    const textElmExpStyle = {
      ...commonTextElmExpStyle,
      fontSize: 20,
      lineHeight: 28,
    };
    const size = 'm';
    const { getByTestId } = render(
      <Avatar participant={testParticipant} size={size} borderColor={borderColor} testID={testID} />,
    );
    const element = getByTestId(testID);
    const styles = mergeStyles(element.props.style);
    const txtElement = getByTestId(textElmTestID);
    const textElmStyle = mergeStyles(txtElement.props.style);
    expect(styles).toMatchObject(expStyles);
    expect(textElmStyle).toMatchObject(textElmExpStyle);
  });
  test('matches style with size:l', () => {
    const expStyles = {
      ...commonExpStyle,
      height: 80,
      width: 80,
      borderRadius: 40,
    };
    const textElmExpStyle = {
      ...commonTextElmExpStyle,
      fontSize: 32,
      lineHeight: 40,
    };
    const size = 'l';
    const { getByTestId } = render(
      <Avatar participant={testParticipant} size={size} borderColor={borderColor} testID={testID} />,
    );
    const element = getByTestId(testID);
    const styles = mergeStyles(element.props.style);
    const txtElement = getByTestId(textElmTestID);
    const textElmStyle = mergeStyles(txtElement.props.style);
    expect(styles).toMatchObject(expStyles);
    expect(textElmStyle).toMatchObject(textElmExpStyle);
  });
  test('matches style for size:m when participant is string type with first&last name only', () => {
    const size = 'm';
    const participant = 'Foo SYDQA';
    const { getByTestId } = render(
      <Avatar participant={participant} size={size} borderColor={borderColor} testID={testID} />,
    );
    const txtElement = getByTestId(textElmTestID);
    expect(txtElement.props.children).toHaveLength(2);
    expect(txtElement.props.children).toMatchObject(['F', 'S']);
  });
  test('matches style for size:l when participant is string type with first&last name only', () => {
    const size = 'l';
    const participant = 'Foo SYDQA';
    const { getByTestId } = render(
      <Avatar participant={participant} size={size} borderColor={borderColor} testID={testID} />,
    );
    const txtElement = getByTestId(textElmTestID);
    expect(txtElement.props.children).toHaveLength(2);
    expect(txtElement.props.children).toMatchObject(['F', 'S']);
  });
  test('matches style for size:l when participant is string type with first&last name only', () => {
    const size = 's';
    const participant = 'Foo SYDQA';
    const { getByTestId } = render(
      <Avatar participant={participant} size={size} borderColor={borderColor} testID={testID} />,
    );
    const txtElement = getByTestId(textElmTestID);
    expect(txtElement.props.children).toHaveLength(2);
    expect(txtElement.props.children).toMatchObject(['F', undefined]);
  });
  test('matches style for size:m when participant is string type with first name only', () => {
    const size = 'm';
    const participant = 'BarSYDQA';
    const { getByTestId } = render(
      <Avatar participant={participant} size={size} borderColor={borderColor} testID={testID} />,
    );
    const txtElement = getByTestId(textElmTestID);
    expect(txtElement.props.children).toHaveLength(2);
    expect(txtElement.props.children).toMatchObject(['B', '']);
  });
  test('matches style for size:s when participant is string type with first name only', () => {
    const size = 's';
    const participant = 'ZooSYDQA';
    const { getByTestId } = render(
      <Avatar participant={participant} size={size} borderColor={borderColor} testID={testID} />,
    );
    const txtElement = getByTestId(textElmTestID);
    expect(txtElement.props.children).toHaveLength(2);
    expect(txtElement.props.children).toMatchObject(['Z', undefined]);
  });
  test('matches style for size:l when participant is Participant type with first&last name only', () => {
    const size = 'm';
    const { getByTestId } = render(
      <Avatar participant={testParticipant} size={size} borderColor={borderColor} testID={testID} />,
    );
    const txtElement = getByTestId(textElmTestID);
    expect(txtElement.props.children).toHaveLength(2);
    expect(txtElement.props.children).toMatchObject(['D', 'S']);
  });
});

describe('Avatar component snapshot', () => {
  test('should match default param configuration', () => {
    const tree = renderer.create(<Avatar participant={confUsr1} testID={testID} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  test('should match for small size and Participant type participant', () => {
    const size = 's';
    const tree = renderer
      .create(<Avatar participant={testParticipant} size={size} borderColor={borderColor} testID={testID} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  test('should match for medium size and string type participant', () => {
    const size = 'm';
    const tree = renderer
      .create(<Avatar participant={confUsr2} size={size} borderColor={borderColor} testID={testID} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
