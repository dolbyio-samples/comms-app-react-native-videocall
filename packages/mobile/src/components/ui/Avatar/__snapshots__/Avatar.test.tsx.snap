// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar component snapshot should match default param configuration 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "borderWidth": 2,
        "display": "flex",
        "justifyContent": "center",
      },
      Object {
        "backgroundColor": "#737D99",
        "borderColor": "white",
        "borderRadius": 40,
        "height": 80,
        "width": 80,
      },
    ]
  }
  testID="testID"
>
  <Text
    style={
      Array [
        Object {
          "letterSpacing": 1.22,
        },
        Object {
          "fontSize": 32,
          "fontWeight": "600",
          "lineHeight": 40,
        },
        false,
        Object {
          "color": "#FFFFFF",
          "fontFamily": "Roboto",
          "textAlign": "left",
        },
      ]
    }
    testID="testID-text"
  >
    F
  </Text>
</View>
`;

exports[`Avatar component snapshot should match for medium size and string type participant 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "borderWidth": 2,
        "display": "flex",
        "justifyContent": "center",
      },
      Object {
        "backgroundColor": "#F3A052",
        "borderColor": "white",
        "borderRadius": 24,
        "height": 48,
        "width": 48,
      },
    ]
  }
  testID="testID"
>
  <Text
    style={
      Array [
        Object {
          "letterSpacing": 1.22,
        },
        Object {
          "fontSize": 20,
          "fontWeight": "600",
          "lineHeight": 28,
        },
        false,
        Object {
          "color": "#FFFFFF",
          "fontFamily": "Roboto",
          "textAlign": "left",
        },
      ]
    }
    testID="testID-text"
  >
    U
    Q
  </Text>
</View>
`;

exports[`Avatar component snapshot should match for small size and Participant type participant 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "borderWidth": 2,
        "display": "flex",
        "justifyContent": "center",
      },
      Object {
        "backgroundColor": "#FAC819",
        "borderColor": "white",
        "borderRadius": 20,
        "height": 40,
        "width": 40,
      },
    ]
  }
  testID="testID"
>
  <Text
    style={
      Array [
        Object {
          "letterSpacing": 1.22,
        },
        Object {
          "fontSize": 16.6667,
          "fontWeight": "600",
          "lineHeight": 23,
        },
        false,
        Object {
          "color": "#FFFFFF",
          "fontFamily": "Roboto",
          "textAlign": "left",
        },
      ]
    }
    testID="testID-text"
  >
    D
  </Text>
</View>
`;
